javascripts/2048.js

```diff
@@ -1,32 +1,400 @@
+// Constants -----------------------------------------------------------
+const MINSTARTINGTILE = 2;
+const MAXSTARTINGTILE = 4;
+const MINBOARDLOCALE = 0; // starting array index
+const MAXBOARDLOCALE = 3; // highest array index
+const BOARDSIZE = 4; // anything less than 4 is valid
+const WINNINGTILE = 2048;
+// Global Variables ------------------------------------------------------
+var board;
+var score;
+var actionOccurred;
+var alreadyWon;
+var gameOver;
+
 $(document).ready(function() {
-  console.log('ready!');
-  $('body').keydown(function(event){
-    var arrow_keys = [37, 38, 39, 40];
-    if(arrow_keys.indexOf(event.which) > -1) {
-      var tile = $('.tile');
-      moveTile(tile, event.which);
-      event.preventDefault();
+  begin();
+
+  $('#newgame').click(function() {
+    clearBoard();
+    begin();
+  });
+
+  function begin() {
+    board = []; // creates an empty board
+    for (i = 0; i < BOARDSIZE; i++) { // fills the board: creates a 2D array
+      board[i] = new Array(BOARDSIZE);
+    }
+
+    // creates two tiles to start with
+    createTile();
+    createTile();
+
+    score = 0; // sets score to 0 for a new game
+    alreadyWon = false;
+    gameOver = false;
+    changeDisplayedScore();
+  }
+
+  $('body').keydown(function(event) {
+    if (!gameOver) {
+      var arrowKeys = [37, 38, 39, 40];
+      if (arrowKeys.indexOf(event.which) > -1) {
+        moveTiles(event.which);
+        matched(event.which);
+        moveTiles(event.which);
+        if (!isBoardFull() && actionOccurred) {
+          createTile();
+        }
+
+        if (!alreadyWon && hasWon()) {
+          $('img').addClass("blastoff").on("animationend", function() { $(this).removeClass("blastoff") });
+          setTimeout (function(){alert("Congratulations, YOU HAVE WOOOOOON!!!")}, 3000);
+          alreadyWon = true;
+        } else if (alreadyWon && hasLost()) {
+          alert("Congrats on winning!\nBut there are no more moves for you to make.\nPlease start a new game.")
+          gameOver = true;
+        } else if (hasLost()) {
+          $('img').addClass("crashandburn").on("animationend", function() { $(this).removeClass("crashandburn") });
+          setTimeout (function(){alert("YOU HAVE FAILED! D:")}, 1000);
+          gameOver = true;
+        }
+        actionOccurred = false;
+        event.preventDefault();
+      }
     }
   })
-})
+});
 
-function moveTile(tile, direction) {
-  var new_tile_value = tile.attr("data-val") * 2;
-  tile.attr("data-val", new_tile_value);
-  tile.text(new_tile_value);
+// tile creation ------------------------------------------------------
+function randomizeValue() {
```

Comment:
Consider:
- is there another name for this function that would communicate more clearly what it does?
- how could you write the function so that it's easy to change the probabilities of generating different tiles?